tool/tsh/common/vnet_common.go

```diff
@@ -120,6 +122,13 @@ func (p *vnetAppProvider) OnNewConnection(ctx context.Context, profileName, leaf
 	return nil
 }
 
+// OnInvalidLocalPort gets called before VNet refuses to handle a connection to a multi-port TCP app
+// because the provided port does not match any of the TCP ports in the app spec.
+func (p *vnetAppProvider) OnInvalidLocalPort(ctx context.Context, profileName, leafClusterName string, routeToApp proto.RouteToApp) {
+	fmt.Printf("%s: Connection refused, port not included in target ports of app %q.\n",
```

Comment:
log instead of print?

Reply:
I think printing here is fine. This is meant to be a user-facing message, like the notification in Connect or like when `tsh proxy` prints instructions.